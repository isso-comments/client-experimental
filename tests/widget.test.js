/* Test the full Isso widet
 * This test might be a bit too heavy to set up, but is quite powerful
 * */


const $ = require('lib/dom');
const app = require('app');
const offset = require('offset');

const fakeThread = require('fixtures/comment-thread');

const mockExtract = jest.fn(() => []);

var issoApp = null;

beforeEach(() => {
  jest.clearAllMocks();
  jest.mock('counter', () => {
    const originalModule = jest.requireActual('counter');
    return {
      ...originalModule,
      extractThreads: mockExtract,
      //Counter: {
      //  count: jest.fn((objs, cb) => cb([4])),
      //  setCounter: originalModule.setCounter,
      //},
    }
  });
  document.body.innerHTML =
    '<div id=isso-thread></div>' +
    '<script src="http://isso.api/js/embed.min.js"'
          + 'data-isso="/"'
          + 'data-isso-id="1"></script>';

  issoApp = new app.App();
  // Mock svg icons (Jest doesn't read `require`-ed files correctly)
  issoApp.template.templateVars["svg"] = {
    'arrow-up': '<svg></svg>',
    'arrow-down': '<svg></svg>'
  };

  // Fake a server response with date header
  // Normally, this would be used to compute offset between server and client
  // (i.e. translating server time to client time, which may be in a different
  // timezone) -> dunno why server doesn't just reply with UTC? Or does it already?
  const fakeDate = new Date('2022-05-05T11:00:00.000Z'); // comment date + 1h
  offset.update(fakeDate);
});

test('Render whole widget', () => {

  let comment = {
    "id": 2,
    "created": 1651744800.0, // = 2022-05-05T10:00:00.000Z
    "mode": 1,
    "text": "<p>A comment with</p>\n<pre><code>code blocks\nNew line: preformatted\n\nDouble newline\n</code></pre>",
    "author": "John",
    "website": "http://website.org",
    "hash": "4505c1eeda98",
    "parent": null,
  }

  // Mock api.config.then(onSuccess(rv), onFailure(err))
  // Return immediately instead of invoking any promise setTimeout funcs
  const fakeConfigThen = jest.fn(
    (onSuccess, onError) => {onSuccess.call(issoApp, {config: {'require-author': false}})}
  );
  jest.spyOn(issoApp.api, 'config')
    .mockImplementationOnce(() => ({then: fakeConfigThen}));
  const resetSpy = jest.spyOn(issoApp.initDone, 'reset')
  const onReadySpy = jest.spyOn(issoApp.initDone, 'onReady')
  const mergeSpy = jest.spyOn(issoApp, 'mergeConfigs')

  // Mock counter API implementation, return count of 4 comments
  const fakeCountThen = jest.fn(
    (onSuccess, onError) => {onSuccess.call(issoApp, [4])}
  );
  jest.spyOn(issoApp.api, 'count')
    .mockImplementationOnce(() => ({then: fakeCountThen}));
  const counterSpy = jest.spyOn(issoApp.counter, 'count');
  //  .mockImplementationOnce((threads, cb) => cb([4]));

  issoApp.initWidget.call(issoApp);

  expect(counterSpy).toHaveBeenCalledTimes(1);

  // initWidget should call initDone.reset() immediately
  expect(resetSpy).toHaveBeenCalledTimes(1);

  expect(fakeConfigThen).toHaveBeenCalledTimes(1);
  // issoApp.mergeConfigs should be invoked once config has been fetched
  expect(mergeSpy)
    .toHaveBeenCalledTimes(1)
  expect(mergeSpy)
    .toHaveBeenCalledWith({config: {'require-author': false}});
  expect(issoApp.config['require-author']).toBe(false);

  expect(onReadySpy).toHaveBeenCalledTimes(1);
  expect(issoApp.initDone.isReady()).toBeTruthy();

  // Somehow doesn't mock, uses original module...
  //expect(mockExtract).toHaveBeenCalledTimes(1);
  //expect(mockExtract).toHaveReturnedWith([]);

  // Mock api.fetch.then(onSuccess(rv), onFailure(err))
  // Return immediately instead of invoking any promise setTimeout funcs
  const fakeFetchThen = jest.fn(
    (onSuccess, onError) => {onSuccess.call(issoApp, fakeThread)}
    //(onSuccess, onError) => {onSuccess.call(issoApp, {replies: [comment]})}
  );
  jest.spyOn(issoApp.api, 'fetch')
    .mockImplementationOnce(() => ({then: fakeFetchThen}));
  const fetchSpy = jest.spyOn(issoApp, 'fetchComments');

  issoApp.fetchComments.call(issoApp);

  expect(fetchSpy).toHaveBeenCalledTimes(1);
  expect(fakeFetchThen).toHaveBeenCalledTimes(1);

  expect(issoApp.issoThread.innerHTML).toMatchSnapshot();
});
