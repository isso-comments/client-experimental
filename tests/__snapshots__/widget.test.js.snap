// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render whole widget 1`] = `
"<h4 class=\\"isso-thread-heading\\">4 Comments</h4><div class=\\"isso-postbox\\"><div class=\\"isso-form-wrapper\\"><div class=\\"isso-textarea-wrapper\\"><div class=\\"isso-textarea isso-placeholder\\" data-text-placeholder=\\"Type Comment Here (at least 3 chars)\\" contenteditable=\\"true\\">Type Comment Here (at least 3 chars)</div><div class=\\"isso-preview\\"><div class=\\"isso-comment\\"><div class=\\"isso-text-wrapper\\"><div class=\\"isso-text\\"></div></div></div></div></div><section class=\\"isso-auth-section\\"><p class=\\"isso-input-wrapper\\"><label for=\\"isso-postbox-author\\">Name (optional)</label><input id=\\"isso-postbox-author\\" type=\\"text\\" name=\\"author\\" placeholder=\\"John Doe\\" value=\\"\\"></p><p class=\\"isso-input-wrapper\\"><label for=\\"isso-postbox-email\\">E-mail (optional)</label><input id=\\"isso-postbox-email\\" type=\\"email\\" name=\\"email\\" placeholder=\\"johndoe@example.com\\" value=\\"\\"></p><p class=\\"isso-input-wrapper\\"><label for=\\"isso-postbox-website\\">Website (optional)</label><input id=\\"isso-postbox-website\\" type=\\"text\\" name=\\"website\\" placeholder=\\"https://example.com\\" value=\\"\\"></p><p class=\\"isso-post-action\\"><input type=\\"submit\\" value=\\"Submit\\"></p><p class=\\"isso-post-action\\"><input type=\\"button\\" name=\\"preview\\" value=\\"Preview\\"></p><p class=\\"isso-post-action\\"><input type=\\"button\\" name=\\"edit\\" value=\\"Edit\\"></p></section><section class=\\"isso-notification-section\\" style=\\"display: none;\\"><label><input type=\\"checkbox\\" name=\\"notification\\">Subscribe to email notification of replies</label></section></div></div><div id=\\"isso-root\\"><div class=\\"isso-comment\\" id=\\"isso-1\\" data-hash=\\"1548cf654b4d\\"><div class=\\"isso-avatar\\"><svg version=\\"1.1\\" viewBox=\\"0 0 48 48\\" preserveAspectRatio=\\"xMinYMin meet\\" shape-rendering=\\"crispEdges\\" data-hash=\\"1548cf654b4d\\"><rect x=\\"0\\" y=\\"0\\" width=\\"56\\" height=\\"56\\" style=\\"fill: #f0f0f0\\"></rect><rect x=\\"4\\" y=\\"12\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #f19670\\"></rect><rect x=\\"36\\" y=\\"12\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #f19670\\"></rect><rect x=\\"4\\" y=\\"28\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #f19670\\"></rect><rect x=\\"36\\" y=\\"28\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #f19670\\"></rect><rect x=\\"12\\" y=\\"12\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #f19670\\"></rect><rect x=\\"28\\" y=\\"12\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #f19670\\"></rect><rect x=\\"12\\" y=\\"28\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #f19670\\"></rect><rect x=\\"28\\" y=\\"28\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #f19670\\"></rect><rect x=\\"12\\" y=\\"36\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #f19670\\"></rect><rect x=\\"28\\" y=\\"36\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #f19670\\"></rect><rect x=\\"20\\" y=\\"12\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #f19670\\"></rect><rect x=\\"20\\" y=\\"36\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #f19670\\"></rect></svg></div><div class=\\"isso-text-wrapper\\"><div class=\\"isso-comment-header\\" role=\\"meta\\"><a class=\\"isso-author\\" href=\\"https://example.org\\" rel=\\"nofollow\\">Sue</a><span class=\\"isso-spacer\\">•</span><a class=\\"isso-permalink\\" href=\\"#isso-1\\"><time title=\\"Thu May 05 2022 17:05:53 GMT+0000 (Coordinated Universal Time)\\" datetime=\\"2022-05-05T17:05:53Z\\">right now</time></a><span class=\\"isso-note\\"></span></div><div class=\\"isso-text\\"><p>Hi John, thank you for this wonderful article!</p>

<p>However, I still struggle with <em>how</em> to apply the <a href=\\"https://example.org/thing\\" rel=\\"nofollow noopener\\">thing you mentioned</a> to my own use case. I've tried using the <code>malleable logarithmic casing</code> you suggested, but somehow my <strong>Turbo encabulator</strong> still experiences side fumbling.</p>

<p>Any further ideas?</p></div><div class=\\"isso-comment-footer\\"><span class=\\"isso-votes\\">7</span><a class=\\"isso-upvote\\" href=\\"#\\"><svg></svg></a><span class=\\"isso-spacer\\">|</span><a class=\\"isso-downvote\\" href=\\"#\\"><svg></svg></a><a class=\\"isso-reply\\" data-textreply=\\"Reply\\" data-textclose=\\"Close\\" href=\\"#\\">Reply</a></div></div><div class=\\"isso-follow-up\\"><div class=\\"isso-comment\\" id=\\"isso-3\\" data-hash=\\"8c2488ea3011\\"><div class=\\"isso-avatar\\"><svg version=\\"1.1\\" viewBox=\\"0 0 48 48\\" preserveAspectRatio=\\"xMinYMin meet\\" shape-rendering=\\"crispEdges\\" data-hash=\\"8c2488ea3011\\"><rect x=\\"0\\" y=\\"0\\" width=\\"56\\" height=\\"56\\" style=\\"fill: #f0f0f0\\"></rect><rect x=\\"4\\" y=\\"4\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #5698c4\\"></rect><rect x=\\"36\\" y=\\"4\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #5698c4\\"></rect><rect x=\\"4\\" y=\\"36\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #5698c4\\"></rect><rect x=\\"36\\" y=\\"36\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #5698c4\\"></rect><rect x=\\"12\\" y=\\"4\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #5698c4\\"></rect><rect x=\\"28\\" y=\\"4\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #5698c4\\"></rect><rect x=\\"20\\" y=\\"28\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #5698c4\\"></rect></svg></div><div class=\\"isso-text-wrapper\\"><div class=\\"isso-comment-header\\" role=\\"meta\\"><span class=\\"isso-author\\">Anonymous</span><span class=\\"isso-spacer\\">•</span><a class=\\"isso-permalink\\" href=\\"#isso-3\\"><time title=\\"Thu May 05 2022 17:08:26 GMT+0000 (Coordinated Universal Time)\\" datetime=\\"2022-05-05T17:08:26Z\\">right now</time></a><span class=\\"isso-note\\"></span></div><div class=\\"isso-text\\"><p>Hey, I'm a bit privacy conscious so I commented anonymously - which works great, thanks!</p>

<p>I struggle with a similar issue as Sue - could it be related to me only using five hydrocoptic marzelvanes?</p></div><div class=\\"isso-comment-footer\\"><span class=\\"isso-votes\\">1</span><a class=\\"isso-upvote\\" href=\\"#\\"><svg></svg></a><span class=\\"isso-spacer\\">|</span><a class=\\"isso-downvote\\" href=\\"#\\"><svg></svg></a><a class=\\"isso-reply\\" data-textreply=\\"Reply\\" data-textclose=\\"Close\\" href=\\"#\\">Reply</a></div></div><div class=\\"isso-follow-up\\"></div></div><div class=\\"isso-comment\\" id=\\"isso-4\\" data-hash=\\"0c7d4cd57a0f\\"><div class=\\"isso-avatar\\"><svg version=\\"1.1\\" viewBox=\\"0 0 48 48\\" preserveAspectRatio=\\"xMinYMin meet\\" shape-rendering=\\"crispEdges\\" data-hash=\\"0c7d4cd57a0f\\"><rect x=\\"0\\" y=\\"0\\" width=\\"56\\" height=\\"56\\" style=\\"fill: #f0f0f0\\"></rect><rect x=\\"4\\" y=\\"12\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #447c69\\"></rect><rect x=\\"36\\" y=\\"12\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #447c69\\"></rect><rect x=\\"4\\" y=\\"28\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #447c69\\"></rect><rect x=\\"36\\" y=\\"28\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #447c69\\"></rect><rect x=\\"4\\" y=\\"36\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #447c69\\"></rect><rect x=\\"36\\" y=\\"36\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #447c69\\"></rect><rect x=\\"12\\" y=\\"4\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #447c69\\"></rect><rect x=\\"28\\" y=\\"4\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #447c69\\"></rect><rect x=\\"12\\" y=\\"12\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #447c69\\"></rect><rect x=\\"28\\" y=\\"12\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #447c69\\"></rect><rect x=\\"12\\" y=\\"28\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #447c69\\"></rect><rect x=\\"28\\" y=\\"28\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #447c69\\"></rect><rect x=\\"20\\" y=\\"36\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #447c69\\"></rect></svg></div><div class=\\"isso-text-wrapper\\"><div class=\\"isso-comment-header\\" role=\\"meta\\"><a class=\\"isso-author\\" href=\\"http://website.org\\" rel=\\"nofollow\\">John</a><span class=\\"isso-spacer\\">•</span><a class=\\"isso-permalink\\" href=\\"#isso-4\\"><time title=\\"Thu May 05 2022 17:14:07 GMT+0000 (Coordinated Universal Time)\\" datetime=\\"2022-05-05T17:14:07Z\\">right now</time></a><span class=\\"isso-note\\"></span></div><div class=\\"isso-text\\"><p>Hi sue, hi anonymous commenter, great to hear you enjoyed my article. Have you ensured that the two spurving bearings are in a direct line with the <code>pentametric fan</code>?</p></div><div class=\\"isso-comment-footer\\"><span class=\\"isso-votes\\">3</span><a class=\\"isso-upvote\\" href=\\"#\\"><svg></svg></a><span class=\\"isso-spacer\\">|</span><a class=\\"isso-downvote\\" href=\\"#\\"><svg></svg></a><a class=\\"isso-reply\\" data-textreply=\\"Reply\\" data-textclose=\\"Close\\" href=\\"#\\">Reply</a></div></div><div class=\\"isso-follow-up\\"></div></div></div></div><div class=\\"isso-comment\\" id=\\"isso-2\\" data-hash=\\"6688cdc5cb0b\\"><div class=\\"isso-avatar\\"><svg version=\\"1.1\\" viewBox=\\"0 0 48 48\\" preserveAspectRatio=\\"xMinYMin meet\\" shape-rendering=\\"crispEdges\\" data-hash=\\"6688cdc5cb0b\\"><rect x=\\"0\\" y=\\"0\\" width=\\"56\\" height=\\"56\\" style=\\"fill: #f0f0f0\\"></rect><rect x=\\"4\\" y=\\"12\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #9163b6\\"></rect><rect x=\\"36\\" y=\\"12\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #9163b6\\"></rect><rect x=\\"4\\" y=\\"20\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #9163b6\\"></rect><rect x=\\"36\\" y=\\"20\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #9163b6\\"></rect><rect x=\\"4\\" y=\\"28\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #9163b6\\"></rect><rect x=\\"36\\" y=\\"28\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #9163b6\\"></rect><rect x=\\"12\\" y=\\"12\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #9163b6\\"></rect><rect x=\\"28\\" y=\\"12\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #9163b6\\"></rect><rect x=\\"12\\" y=\\"28\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #9163b6\\"></rect><rect x=\\"28\\" y=\\"28\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #9163b6\\"></rect><rect x=\\"12\\" y=\\"36\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #9163b6\\"></rect><rect x=\\"28\\" y=\\"36\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #9163b6\\"></rect><rect x=\\"20\\" y=\\"36\\" width=\\"8\\" height=\\"8\\" style=\\"fill: #9163b6\\"></rect></svg></div><div class=\\"isso-text-wrapper\\"><div class=\\"isso-comment-header\\" role=\\"meta\\"><a class=\\"isso-author\\" href=\\"https://example.org\\" rel=\\"nofollow\\">Angelo</a><span class=\\"isso-spacer\\">•</span><a class=\\"isso-permalink\\" href=\\"#isso-2\\"><time title=\\"Thu May 05 2022 17:06:49 GMT+0000 (Coordinated Universal Time)\\" datetime=\\"2022-05-05T17:06:49Z\\">right now</time></a><span class=\\"isso-note\\"></span></div><div class=\\"isso-text\\"><p>Great to see you writing again, John! Looking forward to meeting up again and sharing family pictures.</p></div><div class=\\"isso-comment-footer\\"><span class=\\"isso-votes\\">4</span><a class=\\"isso-upvote\\" href=\\"#\\"><svg></svg></a><span class=\\"isso-spacer\\">|</span><a class=\\"isso-downvote\\" href=\\"#\\"><svg></svg></a><a class=\\"isso-reply\\" data-textreply=\\"Reply\\" data-textclose=\\"Close\\" href=\\"#\\">Reply</a></div></div><div class=\\"isso-follow-up\\"></div></div></div>"
`;
